rust/agama-server/src/agama-web-server.rs

```diff
@@ -32,23 +60,213 @@ struct Cli {
     pub command: Commands,
 }
 
+// check whether the connection uses SSL or not
+async fn is_ssl_stream(stream: &tokio::net::TcpStream) -> bool {
+    // a buffer for reading the first byte from the TCP connection
+    let mut buf = [0u8; 1];
+
+    // peek() receives the data without removing it from the stream,
+    // the data is not consumed, it will be read from the stream again later
+    stream.peek(&mut buf).await.unwrap();
+
+    // SSL3.0/TLS1.x starts with byte 0x16
+    // SSL2 starts with 0x80 (but should not be used as it is considered)
+    // see https://stackoverflow.com/q/3897883
+    // otherwise consider the stream as a plain HTTP stream possibly starting with
+    // "GET ... HTTP/1.1" or "POST ... HTTP/1.1" or a similar line
+    buf[0] == 0x16u8 || buf[0] == 0x80u8
+}
+
+// build a SSL acceptor using a provided SSL certificate or generate a self-signed one
+fn create_ssl_acceptor(cert_file: &String, key_file: &String) -> SslAcceptor {
+    let mut tls_builder = SslAcceptor::mozilla_modern_v5(SslMethod::tls_server()).unwrap();
+
+    if cert_file.is_empty() && key_file.is_empty() {
+        let (cert, key) = agama_dbus_server::cert::create_certificate().unwrap();
+        tracing::info!("Generated self signed certificate: {:#?}", cert);
+        tls_builder.set_private_key(key.as_ref()).unwrap();
+        tls_builder.set_certificate(cert.as_ref()).unwrap();
+
+        // for debugging you might dump the certificate to a file:
+        // use std::io::Write;
+        // let mut cert_file = std::fs::File::create("agama_cert.pem").unwrap();
+        // let mut key_file = std::fs::File::create("agama_key.pem").unwrap();
+        // cert_file.write_all(cert.to_pem().unwrap().as_ref()).unwrap();
+        // key_file.write_all(key.private_key_to_pem_pkcs8().unwrap().as_ref()).unwrap();
+    } else {
+        tracing::info!("Loading PEM certificate: {}", cert_file);
+        tls_builder
+            .set_certificate_file(PathBuf::from(cert_file), SslFiletype::PEM)
+            .unwrap();
+
+        tracing::info!("Loading PEM key: {}", key_file);
+        tls_builder
+            .set_private_key_file(PathBuf::from(key_file), SslFiletype::PEM)
+            .unwrap();
+    }
+
+    // check that the key belongs to the certificate
+    tls_builder.check_private_key().unwrap();
```

Comment:
in general all those unwrap means hard failure. I think for final code we should at least some of them propagate and e.g. allow ot have http only when building it failed.

Reply:
it was also my thought .... may be you're looking (as i was) for things like ```?```, ```is_ok()```, ```ok_or```, ```and_then``` and others (depending whether you're dealing with Option or Result, Option also contains some funny methods for converting into a Result directly)